### PR TITLE
New Tenant CRD with multi Statefulset Support

### PR DESCRIPTION
**Pull request by dvaldivia**
Updated Tenant CRD which supports multiple statefulsets, one per-zone.
Introduces new validations
A folder called `testing` is included. This folder includes a variety of yamls with tenants that will fail validations or tests a specific configuration.

Screen Shot 2020-07-21 at 4 56 18 PM

### CODE REVIEW

File: pkg/controller/cluster/main-controller.go [hunk 1]
```diff
@@ -352,31 +353,32 @@ func (c *Controller) syncHandler(key string) error {
...
 			return nil
 		}
-		return err
+		return nil
 	}
-
 	// Set any required default values and init Global variables
 	nsName := types.NamespacedName{Namespace: namespace, Name: name}
-	mi.EnsureDefaults()
+	mi = mi.EnsureDefaults()
```
Comment (kerneltime) [inline pkg/controller/cluster/main-controller.go:360]: Why the assignment? The function modifies the receiver

Comment (dvaldivia) [reply pkg/controller/cluster/main-controller.go:360]: we return the instance itself from that function. 

There is a bug in the present code where the defaults would get lost out of nowhere, it was caused by the status update, which returns an updated tenant object.

File: pkg/controller/cluster/main-controller.go [hunk 2]
```diff
@@ -352,31 +353,32 @@ func (c *Controller) syncHandler(key string) error {
...
 	if err = mi.Validate(); err != nil {
 		klog.V(2).Infof(err.Error())
 		var err2 error
-		mi, err2 = c.updateTenantStatus(ctx, mi, err.Error(), 0)
-		if err2 != nil {
+		if mi, err2 = c.updateTenantStatus(ctx, mi, err.Error(), 0); err2 != nil {
 			klog.V(2).Infof(err2.Error())
 		}
-		return err
+		return nil
```
Comment (kerneltime) [inline pkg/controller/cluster/main-controller.go:369]: The processNextWorkItem() loops logs and error at log level error. Why skip returning the error? Here and below.

File: pkg/controller/cluster/main-controller.go [hunk 3]
```diff
@@ -387,19 +389,23 @@ func (c *Controller) syncHandler(key string) error {
 		if !(mi.ExternalCert() && mi.ExternalClientCert() && mi.KESExternalCert()) {
 			msg := "Please provide certificate secrets for MinIO and KES, since automatic TLS is disabled"
 			klog.V(2).Infof(msg)
-			return fmt.Errorf(msg)
+			var err2 error
+			if mi, err2 = c.updateTenantStatus(ctx, mi, msg, 0); err2 != nil {
+				klog.V(2).Infof(err2.Error())
+			}
+			return nil
```
Comment (kerneltime) [inline pkg/controller/cluster/main-controller.go:396]: Why not return the error?

File: pkg/controller/cluster/main-controller.go (hunk 2 above)
Comment (dvaldivia) [reply pkg/controller/cluster/main-controller.go:369]: It's causing the Tenants that go into an error state to loop endlessly, if nothing was changed to fix this problem, there's no sense in re-processing the objects. The error is still returned during the zones loop

File: pkg/controller/cluster/main-controller.go (hunk 3 above)
Comment (dvaldivia) [reply pkg/controller/cluster/main-controller.go:396]: Same answer as above

File: hack/update-codegen.sh [hunk 4]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+#
+# Copyright 2020 Google LLC
```
Comment (kerneltime) [inline hack/update-codegen.sh:3]: @dvaldivia I guess this is from another place? Do we want to mix the licenses or rewrite something on our own?

File: hack/update-codegen.sh [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+SCRIPT_ROOT=$(dirname ${BASH_SOURCE})/..
+
+go get -d k8s.io/code-generator/...
+
+# Checkout code-generator to compatible version
+#(cd $GOPATH/src/k8s.io/code-generator && git checkout origin/release-1.14 -B release-1.14)
+
+REPOSITORY=github.com/minio/minio-operator
+$GOPATH/src/k8s.io/code-generator/generate-groups.sh all \
+  $REPOSITORY/pkg/client $REPOSITORY/pkg/apis minio.min.io:v1 \
```
Comment (kerneltime) [inline hack/update-codegen.sh:30]: @dvaldivia This needs updating.

File: hack/update-codegen.sh (hunk 4 above)
Comment (kerneltime) [reply hack/update-codegen.sh:3]: I guess this file was not expected to be checked in.

**Comment by nitisht** [comment]
With zone setup as this

```yaml
zones:
    - servers: 4
      volumesPerServer: 2
      volumeClaimTemplate:
        metadata:
          name: data
        spec:
          accessModes:
            - ReadWriteOnce
          resources:
            requests:
              storage: 1Ti
    - servers: 4
      volumesPerServer: 4
      volumeClaimTemplate:
        metadata:
          name: data
        spec:
          accessModes:
            - ReadWriteOnce
          resources:
            requests:
              storage: 1Ti
```

Operator creates incorrect command line args 

```
https://minio-zone-0-{0...3}.minio-hl.default.svc.cluster.local/export{0...1}
https://minio-zone-1-{0...3}.minio-hl.default.svc.cluster.local/export{0...1}
```

It should be 

```
https://minio-zone-0-{0...3}.minio-hl.default.svc.cluster.local/export{0...1}
https://minio-zone-1-{0...3}.minio-hl.default.svc.cluster.local/export{0...3}
```

**Comment by nitisht** [comment]
Patching an existing Tenant did not add new statefulset / pods

File: pkg/controller/cluster/main-controller.go [hunk 6]
```diff
@@ -735,7 +732,9 @@ func (c *Controller) updateTenantStatus(ctx context.Context, tenant *miniov1.Ten
 	// we must use Update instead of UpdateStatus to update the Status block of the Tenant resource.
 	// UpdateStatus will not allow changes to the Spec of the resource,
 	// which is ideal for ensuring nothing other than resource status has been updated.
-	return c.minioClientSet.MinioV1().Tenants(tenant.Namespace).UpdateStatus(ctx, tenantCopy, opts)
+	t, err := c.minioClientSet.MinioV1().Tenants(tenant.Namespace).UpdateStatus(ctx, tenantCopy, opts)
+	t.EnsureDefaults()
```
[the remark this answers is not in the document]
Comment (dvaldivia) [reply pkg/controller/cluster/main-controller.go:736]: After we call `UpdateStatus` we were getting a new Tenant instance, this Tenant would loose all the defaults previously added, thus leading to bugs, for example, lost serviceName, mountPath and others, this fixes that bug which is present in `2.0.9` as well

File: pkg/resources/statefulsets/minio-statefulset.go [hunk 7]
```diff
@@ -90,7 +90,7 @@ func minioEnvironmentVars(t *miniov1.Tenant) []corev1.EnvVar {
 // Returns the MinIO pods metadata set in configuration.
 // If a user specifies metadata in the spec we return that
 // metadata.
-func minioMetadata(t *miniov1.Tenant) metav1.ObjectMeta {
+func minioMetadata(t *miniov1.Tenant, zone *miniov1.Zone) metav1.ObjectMeta {
```
[the remark this answers is not in the document]
Comment (dvaldivia) [reply pkg/resources/statefulsets/minio-statefulset.go:93]: true, I'll remove it

**Comment by dvaldivia** [comment]
@nitisht both issues are fixed now

File: hack/update-codegen.sh (hunk 5 above)
Comment (dvaldivia) [reply hack/update-codegen.sh:30]: It already references the packages and folders we use in this project, what needs updating?

File: hack/update-codegen.sh (hunk 4 above)
Comment (dvaldivia) [reply hack/update-codegen.sh]: updated

File: pkg/controller/cluster/main-controller.go [hunk 8]
```diff
@@ -429,74 +435,84 @@ func (c *Controller) syncHandler(key string) error {
...
-			ss, err = c.kubeClientSet.AppsV1().StatefulSets(mi.Namespace).Create(ctx, ss, cOpts)
-			if err != nil {
+				if mi, err = c.updateTenantStatus(ctx, mi, statusProvisioningStatefulSet, 0); err != nil {
+					return err
+				}
+				ss = statefulsets.NewForMinIOZone(mi, &zone, hlSvc.Name, c.hostsTemplate)
+				ss, err = c.kubeClientSet.AppsV1().StatefulSets(mi.Namespace).Create(ctx, ss, cOpts)
+				if err != nil {
+					log.Println("ERROR CREATING SS", zone.Name)
+					log.Println(err)
```
[the remark this answers is not in the document]
Comment (dvaldivia) [reply pkg/controller/cluster/main-controller.go]: it was checked in by mistake :P I removed it

**Pull request merged**